Sources/MapboxCoreNavigation/BillingHandler.swift

```diff
@@ -0,0 +1,379 @@
+        if triggerBillingServiceEvents {
+            billingService.triggerBillingEvent(onError: { error in
+                print(error)
+            })
+            billingService.beginBillingSession(for: sessionType, onError: { [weak self] error in
+                self?.failedToBeginBillingSession(with: uuid, with: error)
+            })
+        }
+    }
+
+    /// Stops the billing session identified by the `uuid`.
+    func stopBillingSession(with uuid: UUID) {
+        lock.lock()
+        guard let session = _sessions[uuid] else {
+            lock.unlock(); return
+        }
+        _sessions[uuid] = nil
+
+        let triggerBillingServiceEvents = !_hasSession(with: session.type)
+            && billingService.getSessionStatus(for: session.type) != .stopped
+        lock.unlock()
+
+        if triggerBillingServiceEvents {
+            billingService.stopBillingSession(for: session.type)
+        }
+    }
+ 
+    /// Pauses the billing session identified by the `uuid`.
+    func pauseBillingSession(with uuid: UUID) {
+        lock.lock()
+        guard var session = _sessions[uuid] else {
+            assertionFailure("Trying to pause non-existing session.")
+            lock.unlock(); return
+        }
+        session.isPaused = true
+        _sessions[uuid] = session
+
+
+        let triggerBillingServiceEvent = !_hasSession(with: session.type, isPaused: false)
+            && billingService.getSessionStatus(for: session.type) == .running
+        lock.unlock()
+
+        if triggerBillingServiceEvent {
+            billingService.pauseBillingSession(for: session.type)
+        }
+    }
+    
+    /// Resumes the billing session identified by the `uuid`.
+    func resumeBillingSession(with uuid: UUID) {
+        lock.lock()
+        guard var session = _sessions[uuid] else {
+            assertionFailure("Trying to pause non-existing session.")
+            lock.unlock(); return
+        }
+        session.isPaused = false
+        _sessions[uuid] = session
+        let triggerBillingServiceEvent = billingService.getSessionStatus(for: session.type) == .paused
+        lock.unlock()
+
+        if triggerBillingServiceEvent {
+            billingService.resumeBillingSession(for: session.type) { _ in
+                self.failedToResumeBillingSession(with: uuid)
+            }
+        }
+    }
+
+    private func failedToBeginBillingSession(with uuid: UUID, with error: Error) {
+        lock {
+            _sessions[uuid] = nil
+        }
+    }
+
+    private func failedToResumeBillingSession(with uuid: UUID) {
+        lock.lock()
+        guard let session = _sessions[uuid] else {
+            lock.unlock(); return
+        }
+        _sessions[uuid] = nil
+        lock.unlock()
+        beginBillingSession(for: session.type, uuid: uuid)
+    }
+
+    private func _sessionTypeForRequests() -> SessionType? {
+        for session in _sessions.values {
+            if session.type == .activeGuidance {
```

Comment:
Why is `activeGuidance` preferred over free drive? Could we just pick 'any' session from `sessions`?

Reply:
business rules.